### PR TITLE
Add Couchdb logging receiver

### PR DESCRIPTION
**Pull request by JonathanWamsley**
Added logging receiver for couchdb.

A single receiver that supports 2 log formats as access logs or error logs:
Logs are received from a single source /var/log/couchdb/couchdb.log

Sample log example:
```yaml
{
  "insertId": "152cslufvw01o4",
  "jsonPayload": {
    "message": "admin GET /forbidden 404 ok 1",
    "status_message": "ok",
    "remote_user": "admin",
    "pid": "0.25387.37",
    "path": "/forbidden",
    "host": "127.0.0.1",
    "node": "couchdb"
  },
  "httpRequest": {
    "requestMethod": "GET",
    "status": 404,
    "responseSize": "1",
    "remoteIp": "127.0.0.1",
    "serverIp": "localhost:5984"
  },
  "resource": {
    "type": "gce_instance",
    "labels": {
      "instance_id": "1314100884383110088",
      "project_id": "otel-agent-dev",
      "zone": "us-central1-a"
    }
  },
  "timestamp": "2022-02-10T20:34:32.105337Z",
  "severity": "NOTICE",
  "logName": "projects/otel-agent-dev/logs/couchdb",
  "receiveTimestamp": "2022-02-10T20:34:34.937004216Z"
}
```

### CODE REVIEW

File: apps/couchdb.go [hunk 1]
```diff
@@ -61,3 +62,122 @@ func (r MetricsReceiverCouchdb) Pipelines() []otel.Pipeline {
...
+						},
+					},
+				},
+				{
+					// Format https://github.com/apache/couchdb/blob/main/src/couch_log/src/couch_log_writer_syslog.erl#L72
+					/* Sample line:
+					[error] 2022-01-12T16:53:03.094488Z nonode@nohost emulator -------- Error in process <0.463.0> with exit value:
+					{database_does_not_exist,[{mem3_shards,load_shards_from_db,"_users",[{file,"src/mem3_shards.erl"},{line,399}]},{mem3_shards,load_shards_from_disk,1,[{file,"src/mem3_shards.erl"},{line,374}]},{mem3_shards,load_shards_from_disk,2,[{file,"src/mem3_shards.erl"},{line,403}]},{mem3_shards,for_docid,3,[{file,"src/mem3_shards.erl"},{line,96}]},{fabric_doc_open,go,3,[{file,"src/fabric_doc_open.erl"},{line,39}]},{chttpd_auth_cache,ensure_auth_ddoc_exists,2,[{file,"src/chttpd_auth_cache.erl"},{line,198}]},{chttpd_auth_cache,listen_for_changes,1,[{file,"src/chttpd_auth_cache.erl"},{line,145}]}]}
+					*/
+					Regex: `^\[(?<severity>\w*)\] (?<timestamp>[\d\-\.:TZ]+) (?<node>\w+)@(?<host>[^\s]+) (\w+) [\w-]+ (?<message>[\s\S]*)`,
```
Comment (dehaansa) [inline apps/couchdb.go:97]: Are you certain "node" can only contain letters?
Why are you removing "emulator --------" from the message?

File: apps/couchdb.go [hunk 2]
```diff
@@ -61,3 +62,122 @@ func (r MetricsReceiverCouchdb) Pipelines() []otel.Pipeline {
...
+		Rules: []confgenerator.MultilineRule{
+			{
+				StateName: "start_state",
+				NextState: "cont",
+				Regex:     `^\[\w+\]`,
+			},
+			{
+				StateName: "cont",
+				NextState: "cont",
+				Regex:     `^\[\w+\]`,
```
Comment (dehaansa) [inline apps/couchdb.go:123]: The continuation pattern shouldn't be identical to the start state.

File: apps/couchdb.go [hunk 3]
```diff
@@ -61,3 +62,122 @@ func (r MetricsReceiverCouchdb) Pipelines() []otel.Pipeline {
...
+}
+
+func (p LoggingProcessorCouchdbError) Components(tag string, uid string) []fluentbit.Component {
+	c := confgenerator.LoggingProcessorParseMultilineRegex{
+		LoggingProcessorParseRegexComplex: confgenerator.LoggingProcessorParseRegexComplex{
+			Parsers: []confgenerator.RegexParser{
+				{
+					// Format https://github.com/apache/couchdb/blob/main/src/couch_log/src/couch_log_writer_syslog.erl#L72
+					// Sample line: [notice] 2021-12-02T23:36:42.555157Z nonode@nohost <0.17165.1> a5f585a0d3 localhost:5984 127.0.0.1 otelu PUT /oteld 201 ok 16
+					Regex: `^\[(?<severity>\w*)\] (?<timestamp>[\d\-\.:TZ]+) (?<node>\w+)@(?<host>[^\s]+) \<(?<pid>[^ ]*)\> [\w-]+ (?<http_request_serverIp>[^ ]*) (?<http_request_remoteIp>[^ ]*) (?<remote_user>[^ ]*) (?<http_request_requestMethod>[^ ]*) (?<path>[^ ]*) (?<http_request_status>[^ ]*) (?<message>[^ ]*) (?<http_request_responseSize>[\d]*)$`,
```
Comment (dehaansa) [inline apps/couchdb.go:81]: I don't think what you've captured here as "message" would be useful. Maybe capture remote_user through the end as message. You can still capture subfields inside (nested capture groups), but this'll leave a useful message in the logs explorer as the "message".

File: apps/couchdb.go [hunk 4]
```diff
@@ -61,3 +62,122 @@ func (r MetricsReceiverCouchdb) Pipelines() []otel.Pipeline {
 func init() {
 	confgenerator.MetricsReceiverTypes.RegisterType(func() confgenerator.Component { return &MetricsReceiverCouchdb{} })
 }
+
+type LoggingProcessorCouchdbError struct {
```
Comment (dehaansa) [inline apps/couchdb.go:66]: Struct shouldn't be named "error"

File: apps/couchdb.go [hunk 5]
```diff
@@ -61,3 +62,122 @@ func (r MetricsReceiverCouchdb) Pipelines() []otel.Pipeline {
...
+
+func (r LoggingReceiverCouchdbError) Components(tag string) []fluentbit.Component {
+	if len(r.IncludePaths) == 0 {
+		r.IncludePaths = []string{
+			// Default log file
+			"/var/log/couchdb/couchdb.log",
+		}
+	}
+	c := r.LoggingReceiverFilesMixin.Components(tag)
+	c = append(c, r.LoggingProcessorCouchdbError.Components(tag, "couchdb_error")...)
```
Comment (dehaansa) [inline apps/couchdb.go:176]: fix to be "couchdb" like the type named above.

File: apps/couchdb.go (hunk 2 above)
Comment (JonathanWamsley) [reply apps/couchdb.go]: Changed the continuation pattern

File: apps/couchdb.go (hunk 3 above)
Comment (JonathanWamsley) [reply apps/couchdb.go]: updated message to capture remote user to end. Renamed old message to response_message.

File: apps/couchdb.go (hunk 4 above)
Comment (JonathanWamsley) [reply apps/couchdb.go]: fixed

File: apps/couchdb.go (hunk 5 above)
Comment (JonathanWamsley) [reply apps/couchdb.go]: fixed

File: apps/couchdb.go [hunk 6]
```diff
@@ -61,3 +62,114 @@ func (r MetricsReceiverCouchdb) Pipelines() []otel.Pipeline {
...
+					*/
+					Regex: `^\[(?<level>\w*)\] (?<timestamp>[\d\-\.:TZ]+) (?<node>\S+)@(?<host>[^\s]+) (?<message>[\s\S]*(\<(?<pid>[^>]+)\>)[\s\S]*)`,
+					Parser: confgenerator.ParserShared{
+						TimeKey:    "timestamp",
+						TimeFormat: "%Y-%m-%dT%H:%M:%S.%L%z",
+					},
+				},
+			},
+		},
+		Rules: []confgenerator.MultilineRule{
```
Comment (dehaansa) [inline apps/couchdb.go:106]: Move multiline details onto input (see https://github.com/GoogleCloudPlatform/ops-agent/pull/399 for examples)

File: integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    httpRequest.remoteIp: "127.0.0.1"
+    httpRequest.requestMethod: "GET"
+    httpRequest.responseSize: "1"
+    httpRequest.serverIp: "localhost:5984"
+    httpRequest.status: 404
+    jsonPayload.host: "127.0.0.1"
+    jsonPayload.message: "admin GET /forbidden 404 ok 1"
+    jsonPayload.node: "couchdb"
+    jsonPayload.path: "/forbidden"
+    jsonPayload.pid: "0.25387.37"
```
Comment (dehaansa) [inline integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml:13]: This is not likely to be consistent between test runs, probably do not try to expect PID at all.

File: integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+log_entries:
+- log_name: "couchdb"
+  field_matchers:
+    httpRequest.remoteIp: "127.0.0.1"
+    httpRequest.requestMethod: "GET"
+    httpRequest.responseSize: "1"
+    httpRequest.serverIp: "localhost:5984"
+    httpRequest.status: 404
```
Comment (dehaansa) [inline integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml:8]: Last time I tried these field matchers only supported strings, might have to remove this for kokoro to work.

File: integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml (hunk 7 above)
Comment (JonathanWamsley) [reply integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml]: updated

File: integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml (hunk 8 above)
Comment (JonathanWamsley) [reply integration_test/third_party_apps_data/applications/couchdb/expected_logs.yaml]: updated

**Comment by qingling128** [comment]
The previous integration test run on commit `[83bf0f8](https://github.com/GoogleCloudPlatform/ops-agent/pull/394/commits/83bf0f80ed2e6bc2ec95118b63e2e1acc97a85fc)` only failed for sles 12:

image

```
        + sudo zypper --non-interactive --no-gpg-checks ref
        Error retrieving metadata for 'SLE-Module-Adv-Systems-Management12-Pool':
        Not ready to read within timeout.
        Skipping repository 'SLE-Module-Adv-Systems-Management12-Pool' because of the above error.
        Some of the repositories have not been refreshed because of an error.
2022/02/15 10:32:17 Attempt 4 of sles-12 test of hbase finished with err=<nil>, retryable=false
2022/02/15 10:34:17 Attempt 4 of sles-12 test of memcached finished with err=<nil>, retryable=false
2022/02/15 10:34:28 Attempt 4 of sles-12 test of hadoop finished with err=<nil>, retryable=false
2022/02/15 10:34:45 Attempt 4 of sles-12 test of zookeeper finished with err=<nil>, retryable=false
2022/02/15 10:38:23 Attempt 4 of sles-12 test of elasticsearch finished with err=<nil>, retryable=false
2022/02/15 10:39:26 Attempt 4 of sles-12 test of tomcat finished with err=<nil>, retryable=false
--- FAIL: TestThirdPartyApps (0.00s)
    --- FAIL: TestThirdPartyApps/sles-12 (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/wildfly (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/jvm (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/couchdb (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/cassandra (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/nginx (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/mysql (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/apache (0.00s)
        --- SKIP: TestThirdPartyApps/sles-12/redis (0.00s)
        --- FAIL: TestThirdPartyApps/sles-12/elasticsearch#01 (1293.23s)
        --- PASS: TestThirdPartyApps/sles-12/mongodb (1293.24s)
        --- FAIL: TestThirdPartyApps/sles-12/activemq (1310.84s)
        --- FAIL: TestThirdPartyApps/sles-12/solr (1321.11s)
        --- PASS: TestThirdPartyApps/sles-12/hbase (1346.48s)
        --- PASS: TestThirdPartyApps/sles-12/kafka (1552.88s)
        --- FAIL: TestThirdPartyApps/sles-12/postgresql (1607.29s)
        --- PASS: TestThirdPartyApps/sles-12/hbase#01 (1656.71s)
        --- PASS: TestThirdPartyApps/sles-12/memcached (1777.33s)
        --- PASS: TestThirdPartyApps/sles-12/hadoop (1790.50s)
        --- PASS: TestThirdPartyApps/sles-12/zookeeper (1802.26s)
        --- PASS: TestThirdPartyApps/sles-12/elasticsearch (2027.83s)
        --- PASS: TestThirdPartyApps/sles-12/tomcat (2094.11s)
```

And Couchdb is not supported on SLES 12.

So we are good integration tests wise. Additional syncing / rebasing off master does not need to trigger integration tests again, unless the change is significant.

**Review by qingling128 — COMMENTED** [review]
A minor thing with the submodule:

image

Collectd has been removed recently.

**Comment by qingling128** [comment]
Looks like the `collectd` submodule is still there. Perhaps just cherry pick https://github.com/GoogleCloudPlatform/ops-agent/pull/400, or `rm -rf submodules/collectd`?

**Review by qingling128 — APPROVED** [review]
LGTM modulo the `collectd` submodule is fixed.

**Pull request merged**